tests/ui/rust-2018/remove-extern-crate.fixed

```diff
@@ -6,6 +6,7 @@
 
 #![warn(rust_2018_idioms)]
 #![allow(dropping_copy_types)]
+#![allow(unused_imports)]
```

Comment:
> Why is this required? (Also, if it's indeed required, then it's better to put `allow` on the specific import, rather than on the whole test.)

If I put allow unused_imports to specific import, lint UNUSED_EXTERN_CRATES will be invalid, and no error will be reported or fixed. The repair strategy of UNUSED_EXTERN_CRATES is to replace `extern crate crate_name` with `use crate_name`, which will cause the fixed file to continue report redundant errors.

Reply:
This part

>repair strategy of UNUSED_EXTERN_CRATES is to replace `extern crate crate_name` with `use crate_name`, which will cause the fixed file to continue report redundant imports errors.

is clear, but I don't understand this part

>If I put allow unused_imports to specific import, lint UNUSED_EXTERN_CRATES will be invalid, and no error will be reported or fixed.

Why is `unused_extern_crates` affected by `unused_imports`?

Reply:
> This part
> 
> > repair strategy of UNUSED_EXTERN_CRATES is to replace `extern crate crate_name` with `use crate_name`, which will cause the fixed file to continue report redundant imports errors.
> 
> is clear, but I don't understand this part
> 
> > If I put allow unused_imports to specific import, lint UNUSED_EXTERN_CRATES will be invalid, and no error will be reported or fixed.
> 
> Why is `unused_extern_crates` affected by `unused_imports`?

It's a bit strange indeed, I think I need to spend some time understanding UNUSED_EXTERN_CRATES for the specific reason. This is log when I add all like:

```rust
#[allow(unused_imports)]
pub extern crate core; //~ WARNING `extern crate` is not idiomatic
```

```bash
---- [ui] tests\ui\rust-2018\remove-extern-crate.rs stdout ----
$DIR\remove-extern-crate.rs
$DIR\remove-extern-crate.rs
diff of stderr:

11         |         ^^^^^^^^^^^^^^^^
12         = note: `#[warn(unused_extern_crates)]` implied by `#[warn(rust_2018_idioms)]`
13
-       warning: `extern crate` is not idiomatic in the new edition
-         --> $DIR/remove-extern-crate.rs:35:5
-          |
-       LL |     extern crate core;
-          |     ^^^^^^^^^^^^^^^^^^
-          |
-       help: convert it to a `use`
-          |
-       LL |     use core;
-          |     ~~~
-
-       warning: `extern crate` is not idiomatic in the new edition
-         --> $DIR/remove-extern-crate.rs:45:5
-          |
-       LL |     pub extern crate core;
-          |     ^^^^^^^^^^^^^^^^^^^^^^
-          |
-       help: convert it to a `use`
-          |
-       LL |     pub use core;
-          |         ~~~
-
-       warning: 3 warnings emitted
+       warning: 1 warning emitted
37
38


The actual stderr differed from the expected stderr.
Actual stderr saved to D:\source\rust\rust\build\x86_64-pc-windows-msvc\test\ui\rust-2018\remove-extern-crate\remove-extern-crate.stderr
diff of fixed:

6
7       #![warn(rust_2018_idioms)]
8       #![allow(dropping_copy_types)]
-       #![allow(unused_imports)]
10
+
11       //~ WARNING unused extern crate
12      // Shouldn't suggest changing to `use`, as `another_name`
13      // would no longer be added to the prelude which could cause

32      }
33
34      mod another {
-           use core; //~ WARNING `extern crate` is not idiomatic
+           #[allow(unused_imports)]
+           extern crate core; //~ WARNING `extern crate` is not idiomatic
36          use remove_extern_crate;
37
38          pub fn foo() {

42      }
43
44      mod with_visibility {
-           pub use core; //~ WARNING `extern crate` is not idiomatic
+           #[allow(unused_imports)]
+           pub extern crate core; //~ WARNING `extern crate` is not idiomatic
46
47          pub fn foo() {
48              core::mem::drop(4);


The actual fixed differed from the expected fixed.
Actual fixed saved to D:\source\rust\rust\build\x86_64-pc-windows-msvc\test\ui\rust-2018\remove-extern-crate\remove-extern-crate.fixed
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args rust-2018\remove-extern-crate.rs`



```

Reply:
> This part
> 
> > repair strategy of UNUSED_EXTERN_CRATES is to replace `extern crate crate_name` with `use crate_name`, which will cause the fixed file to continue report redundant imports errors.
> 
> is clear, but I don't understand this part
> 
> > If I put allow unused_imports to specific import, lint UNUSED_EXTERN_CRATES will be invalid, and no error will be reported or fixed.
> 
> Why is `unused_extern_crates` affected by `unused_imports`?

I find it. 
```rust
In check_unused.rs: 
// If the extern crate has any attributes, they may have funky
// semantics we can't faithfully represent using `use` (most
// notably `#[macro_use]`). Ignore it.
if extern_crate.has_attrs {
    debug!("extern_crate4:{:?}", extern_crate);
    continue;
}
```

Reply:
Ah, I see, let's keep it as is then.